// payme-merchant-integrator/dist/controller/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MainController = void 0;
const app_1 = require("../app");
const errors_1 = require("../errors");
const payme_dto_1 = require("../validators/dtos/payme.dto");
const validate_1 = require("../validators/validate");
class MainController {
    static params;
    static TransactionModel;
    static CounterModel;
    static async handler(request, TransactionModel, CounterModel, params) {
        // initialize model
        this.params = params;
        this.TransactionModel = TransactionModel;
        this.CounterModel = CounterModel;
        const body = await (0, validate_1.validateIt)(request.body, payme_dto_1.PaymeBodyDto, payme_dto_1.paymeGroups);
        return await this.switchMethod(body);
    }
    static async switchMethod(body) {
        switch (body.method) {
            case payme_dto_1.PaymeMethods.CheckPerformTransaction: {
                return await this.CheckPerformTransaction(body.params);
            }
            case payme_dto_1.PaymeMethods.CreateTransaction: {
                return await this.CreateTransaction(body.params);
            }
            case payme_dto_1.PaymeMethods.PerformTransaction: {
                return await this.PerformTransaction(body.params);
            }
            case payme_dto_1.PaymeMethods.CancelTransaction: {
                return await this.CancelTransaction(body.params);
            }
            case payme_dto_1.PaymeMethods.CheckTransaction: {
                return await this.CheckTransaction(body.params);
            }
            case payme_dto_1.PaymeMethods.GetStatement: {
                return this.GetStatement();
            }
        }
    }
    static validatePaymeAmount(amount) {
        amount /= 100;
        return amount;
    }
    static async CheckPerformTransaction(data) {
        const params = await (0, validate_1.validateIt)(data, payme_dto_1.PaymeParamsDto, [payme_dto_1.PaymeMethods.CheckPerformTransaction]);
        params.amount = this.validatePaymeAmount(params.amount);
        const result = await this.params.isAccountExist(params.account);
        if (!result) {
            throw errors_1.PaymeErrors.InvalidRequest('Account not found');
        }
        if (this.params.type == app_1.PaymeIntegratorType.ONE_TIME) {
            const cost = await this.params.getPayingCost(params.account);
            if (cost != params.amount) {
                throw errors_1.PaymeErrors.InvalidAmount();
            }
        }
        return {
            allow: true
        };
    }
    static async CreateTransaction(data) {
        const params = await (0, validate_1.validateIt)(data, payme_dto_1.PaymeParamsDto, [payme_dto_1.PaymeMethods.CreateTransaction]);
        params.amount = this.validatePaymeAmount(params.amount);
        const result = await this.params.isAccountExist(params.account);
        if (!result) {
            throw errors_1.PaymeErrors.InvalidRequest('Account not found');
        }
        if (this.params.type == app_1.PaymeIntegratorType.ONE_TIME) {
            const cost = await this.params.getPayingCost(params.account);
            if (cost != params.amount) {
                throw errors_1.PaymeErrors.InvalidAmount();
            }
        }
        let tra = await this.TransactionModel.findOne({ id: params.id });
        if (tra) {
            return {
                create_time: tra.create_time,
                transaction: tra.transaction,
                state: tra.state
            };
        }
        if (this.params.type == app_1.PaymeIntegratorType.ONE_TIME) {
            const exist = await this.TransactionModel.findOne({ account: params.account });
            if (exist) {
                if (exist.state > 0) {
                    throw errors_1.PaymeErrors.InvalidRequest('Transaction pending');
                }
                await this.TransactionModel.updateOne({
                    _id: exist._id
                }, {
                    $set: {
                        ...params,
                        create_time: data.time,
                        state: 1,
                    }
                });
                tra = await this.TransactionModel.findById(exist._id);
                return {
                    create_time: tra.create_time,
                    transaction: tra.transaction,
                    state: tra.state
                };
            }
        }
        try {
            const { value: transaction } = await this.CounterModel.findOneAndUpdate({ name: 'payme_transactions' }, { $inc: { value: 1 } }, { new: true, upsert: true });
            const tra_params = {
                ...params,
                create_time: data.time,
                state: 1,
                transaction
            };
            const { _id: saved_id } = await new this.TransactionModel(tra_params).save();
            tra = await this.TransactionModel.findById(saved_id);
        }
        catch (error) {
            if (error.name == 'MongoError' && error.code == 11000) {
                const exist = await this.TransactionModel.findOne({ account: params.account });
                if (!exist) {
                    throw errors_1.PaymeErrors.InvalidRequest('Exists smth');
                }
                return {
                    create_time: exist.create_time,
                    transaction: exist.transaction,
                    state: exist.state
                };
            }
            throw errors_1.PaymeErrors.InvalidRequest('Exists');
        }
        return {
            create_time: tra.create_time,
            transaction: tra.transaction,
            state: tra.state
        };
    }
    static async PerformTransaction(data) {
        const params = await (0, validate_1.validateIt)(data, payme_dto_1.PaymeParamsDto, [payme_dto_1.PaymeMethods.PerformTransaction]);
        const tra = await this.TransactionModel.findOne({
            id: params.id
        });
        if (!tra) {
            throw errors_1.PaymeErrors.TransactionNotFound();
        }
        if (tra.state < 0) {
            throw errors_1.PaymeErrors.UnablePerform();
        }
        if (tra.perform_time) {
            return {
                transaction: tra.transaction,
                perform_time: tra.perform_time,
                state: tra.state
            };
        }
        const time = new Date().getTime();
        // pay here first
        await this.params.markAsPaid(tra.account, tra.amount);
        //
        tra.state = 2;
        tra.perform_time = time;
        await this.TransactionModel.updateOne({ _id: tra._id }, {
            $set: {
                perform_time: time,
                state: 2
            }
        });
        return {
            transaction: tra.transaction,
            perform_time: tra.perform_time,
            state: tra.state
        };
    }
    static async CancelTransaction(data) {
        const params = await (0, validate_1.validateIt)(data, payme_dto_1.PaymeParamsDto, [payme_dto_1.PaymeMethods.CancelTransaction]);
        const tra = await this.TransactionModel.findOne({ id: params.id });
        if (!tra) {
            throw errors_1.PaymeErrors.TransactionNotFound();
        }
        const can_cancel = await this.params.canCancel(tra.account);
        if (!can_cancel && tra.state == 2) {
            throw errors_1.PaymeErrors.OrderIsCompleted();
        }
        const cancel_params = {};
        if (!tra.cancel_time) {
            const cancel_time = new Date().getTime();
            cancel_params.cancel_time = cancel_time;
            tra.cancel_time = cancel_time;
        }
        if (!tra.reason) {
            tra.reason = params.reason;
            cancel_params.reason = params.reason;
        }
        if (tra.state > 0) {
            cancel_params.state = tra.state * (-1);
            tra.state *= (-1);
        }
        await this.TransactionModel.updateOne({ _id: tra._id }, { $set: cancel_params });
        await this.params.markAsCancel(tra.account);
        return {
            transaction: tra.transaction,
            cancel_time: tra.cancel_time,
            state: tra.state
        };
    }
    static async CheckTransaction(data) {
        const params = await (0, validate_1.validateIt)(data, payme_dto_1.PaymeParamsDto, [payme_dto_1.PaymeMethods.CheckTransaction]);
        const tra = await this.TransactionModel.findOne({ id: params.id });
        if (!tra) {
            throw errors_1.PaymeErrors.TransactionNotFound();
        }
        return {
            create_time: tra.create_time,
            perform_time: tra.perform_time,
            cancel_time: tra.cancel_time,
            transaction: tra.transaction,
            state: tra.state,
            reason: tra.reason
        };
    }
    static GetStatement() {
        return {
            transactions: []
        };
    }
}
exports.MainController = MainController;
