// payme-merchant-integrator/dist/validators/dtos/payme.dto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymeBodyDto = exports.PaymeParamsDto = exports.paymeGroups = exports.PaymeMethods = void 0;
const tslib_1 = require("tslib");
const class_validator_1 = require("class-validator");
var PaymeMethods;
(function (PaymeMethods) {
    PaymeMethods["CheckPerformTransaction"] = "CheckPerformTransaction";
    PaymeMethods["CreateTransaction"] = "CreateTransaction";
    PaymeMethods["PerformTransaction"] = "PerformTransaction";
    PaymeMethods["CancelTransaction"] = "CancelTransaction";
    PaymeMethods["CheckTransaction"] = "CheckTransaction";
    PaymeMethods["GetStatement"] = "GetStatement";
})(PaymeMethods = exports.PaymeMethods || (exports.PaymeMethods = {}));
exports.paymeGroups = [
    PaymeMethods.CheckPerformTransaction,
    PaymeMethods.CreateTransaction,
    PaymeMethods.PerformTransaction,
    PaymeMethods.CancelTransaction,
    PaymeMethods.CheckPerformTransaction,
    PaymeMethods.GetStatement
];
class PaymeParamsDto {
    id;
    time;
    amount;
    from;
    to;
    reason;
    account;
}
tslib_1.__decorate([
    (0, class_validator_1.IsString)({
        groups: [
            PaymeMethods.CreateTransaction,
            PaymeMethods.PerformTransaction,
            PaymeMethods.CancelTransaction,
            PaymeMethods.CheckTransaction
        ]
    }),
    tslib_1.__metadata("design:type", String)
], PaymeParamsDto.prototype, "id", void 0);
tslib_1.__decorate([
    (0, class_validator_1.IsNumber)({}, {
        groups: [
            PaymeMethods.CreateTransaction
        ]
    }),
    tslib_1.__metadata("design:type", String)
], PaymeParamsDto.prototype, "time", void 0);
tslib_1.__decorate([
    (0, class_validator_1.IsNumber)({}, {
        groups: [
            PaymeMethods.CheckPerformTransaction,
            PaymeMethods.CreateTransaction
        ]
    }),
    tslib_1.__metadata("design:type", Number)
], PaymeParamsDto.prototype, "amount", void 0);
tslib_1.__decorate([
    (0, class_validator_1.IsNumber)({}, {
        groups: [
            PaymeMethods.GetStatement
        ]
    }),
    tslib_1.__metadata("design:type", Number)
], PaymeParamsDto.prototype, "from", void 0);
tslib_1.__decorate([
    (0, class_validator_1.IsNumber)({}, {
        groups: [
            PaymeMethods.GetStatement
        ]
    }),
    tslib_1.__metadata("design:type", Number)
], PaymeParamsDto.prototype, "to", void 0);
tslib_1.__decorate([
    (0, class_validator_1.IsNumber)({}, {
        groups: [
            PaymeMethods.CancelTransaction
        ]
    }),
    tslib_1.__metadata("design:type", Number)
], PaymeParamsDto.prototype, "reason", void 0);
tslib_1.__decorate([
    (0, class_validator_1.IsObject)({
        groups: [
            PaymeMethods.CheckPerformTransaction,
            PaymeMethods.CreateTransaction
        ]
    }),
    tslib_1.__metadata("design:type", Object)
], PaymeParamsDto.prototype, "account", void 0);
exports.PaymeParamsDto = PaymeParamsDto;
class PaymeBodyDto {
    method;
    params;
}
tslib_1.__decorate([
    (0, class_validator_1.IsEnum)(PaymeMethods, { groups: exports.paymeGroups }),
    tslib_1.__metadata("design:type", String)
], PaymeBodyDto.prototype, "method", void 0);
tslib_1.__decorate([
    (0, class_validator_1.IsNotEmpty)({ groups: exports.paymeGroups }),
    (0, class_validator_1.IsObject)({ groups: exports.paymeGroups }),
    tslib_1.__metadata("design:type", Object)
], PaymeBodyDto.prototype, "params", void 0);
exports.PaymeBodyDto = PaymeBodyDto;
