// payme-merchant-integrator/dist/app.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymeIntegrator = exports.PaymeIntegratorType = void 0;
const controller_1 = require("./controller");
const connectToDb_1 = require("./db/connectToDb");
const counter_1 = require("./db/models/counter");
const transaction_1 = require("./db/models/transaction");
const errors_1 = require("./errors");
const middleware_1 = require("./middleware");
var PaymeIntegratorType;
(function (PaymeIntegratorType) {
    PaymeIntegratorType["CUMULATIVE"] = "cumulative";
    PaymeIntegratorType["ONE_TIME"] = "one-time";
})(PaymeIntegratorType = exports.PaymeIntegratorType || (exports.PaymeIntegratorType = {}));
class PaymeIntegrator {
    integratorOptions;
    TransactionModel;
    CounterModel;
    constructor(integratorOptions) {
        this.integratorOptions = integratorOptions;
        if (this.integratorOptions.type == PaymeIntegratorType.ONE_TIME &&
            !this.integratorOptions.getPayingCost) {
            throw Error('getPayingCost is required');
        }
        this.initialize();
    }
    async initialize() {
        try {
            (0, connectToDb_1.connectToDb)(this.integratorOptions.db_str);
            this.TransactionModel = (0, transaction_1.getTransactionModel)(this.integratorOptions.collection, this.integratorOptions.type == PaymeIntegratorType.ONE_TIME);
            this.CounterModel = (0, counter_1.getCounterModel)('payme_counters');
        }
        catch (error) {
            console.error(error);
        }
    }
    async authenticate(request, reply, done = () => { }) {
        try {
            await (0, middleware_1.paymeAuthentication)(request, this.integratorOptions.password);
            done();
        }
        catch (error) {
            if (error instanceof errors_1.PaymeErrors) {
                return reply.send(error);
            }
            return reply.send(errors_1.PaymeErrors.InvalidRequest(error.message));
        }
    }
    async handler(request, reply) {
        try {
            const result = await controller_1.MainController.handler(request, this.TransactionModel, this.CounterModel, this.integratorOptions);
            return reply.send({ result });
        }
        catch (error) {
            if (error instanceof errors_1.PaymeErrors) {
                return reply.send(error);
            }
            reply.send(errors_1.PaymeErrors.InvalidRequest(error.message));
        }
    }
}
exports.PaymeIntegrator = PaymeIntegrator;
