// payme-merchant-integrator/dist/db/models/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTransactionModel = void 0;
const mongoose_1 = require("mongoose");
const TransactionSchema = {
    id: { type: String },
    time: { type: Number },
    amount: { type: Number },
    account: { type: Object, _id: false },
    create_time: { type: Number },
    perform_time: { type: Number, default: 0 },
    cancel_time: { type: Number, default: 0 },
    state: { type: Number },
    reason: { type: Number, default: null },
    transaction: { type: String },
};
const getTransactionModel = (collection = 'transactions', account_index = false) => {
    const schema = new mongoose_1.Schema(TransactionSchema, { collection });
    schema.index({
        id: 1
    }, {
        name: "payme_business_index",
        unique: true,
        background: true,
    });
    if (account_index) {
        schema.index({
            account: 1
        }, {
            name: "account_index",
            unique: true,
            background: true,
        });
    }
    const TransactionModel = (0, mongoose_1.model)('Transaction', schema);
    if (!account_index) {
        TransactionModel.collection.dropIndex('account_index', (error) => { if (error) {
            console.log(error);
        } });
    }
    return TransactionModel;
};
exports.getTransactionModel = getTransactionModel;
